SS14.Launcher/Api/HubApi.cs

```diff
@@ -1,36 +1,121 @@
+    /// <summary>
+    /// Get a list of servers on all hubs. Returns the list of servers as well as a boolean stating whether all fetches
+    /// succeeded.
+    /// </summary>
+    public async Task<(bool AllSucceeded, List<HubServerListEntry> Entries)> GetServers(CancellationToken cancel)
     {
-        return await _http.GetFromJsonAsync<ServerListEntry[]>(
-            ConfigConstants.HubUrl + "api/servers",
-            cancel) ?? throw new JsonException("Server list is null!");
+        var entries = new List<HubServerListEntry>();
+        var requests = new List<(Task<ServerListEntry[]?> Request, Hub Hub)>();
+        var allSucceeded = true;
+
+        // Queue requests
+        foreach (var hub in _dataManager.Hubs)
+        {
+            // Sanity check, this should be enforced with code
+            if (!hub.Address.AbsoluteUri.EndsWith('/'))
+                throw new Exception("URI doesn't have trailing slash");
+
+            requests.Add((_http.GetFromJsonAsync<ServerListEntry[]>(new Uri(hub.Address, "api/servers"), cancel), hub));
+        }
+
+        // Await all requests
+        var tasks = Task.WhenAll(requests.Select(t => t.Request));
+        try
+        {
+            await tasks;
+        }
+        catch (Exception e) when (e is HttpRequestException or JsonException)
+        {
+            if (tasks.Exception?.InnerExceptions is { } inner)
+            {
+                foreach (var ex in inner)
+                {
+                    Log.Warning("Failed fetching servers from a hub: {Message}", ex.Message);
+                }
+            }
+
+            allSucceeded = false;
+        }
+
+        // Process responses
+        foreach (var (request, hub) in requests.OrderBy(x => x.Hub.Priority))
+        {
+            try
+            {
+                var response = await request;
+
+                if (response == null)
+                {
+                    allSucceeded = false;
+                    continue;
+                }
+
+                // Remove duplicate servers
+                // This only removes
+                var deduped = response.Where(e => !entries
+                    .Select(x => x.Address)
+                    .Contains(e.Address)
+                    ).ToArray();
+
+                foreach (var dupe in response.Except(deduped))
+                {
+                    Log.Debug("Removed duplicate server {Address}", dupe.Address);
+                }
+
+                entries.AddRange(deduped.Select(s =>
+                    new HubServerListEntry(s.Address, hub.Address.AbsoluteUri, s.StatusData)));
+            }
+            catch
+            {
+                // Handled by WhenAll
+            }
+        }
+
+        return (allSucceeded, entries);
     }
 
-    public async Task<ServerInfo> GetServerInfo(string serverAddress, CancellationToken cancel)
+
+    public async Task<ServerInfo> GetServerInfo(ServerStatusData statusData, CancellationToken cancel)
```

Comment:
Same thing here. This should take two URIs rather than being knowledgeable of `ServerStatusData`.